protostar/cli/cli_app.py

```diff
@@ -1,16 +1,20 @@
-from typing import Any, List, Optional
+from typing import Any, Generic, Optional, TypeVar
 
-from protostar.cli.command import Command
+from .arg_type import ArgTypeName
+from .argument import Argument
+from .command import Command
 
+ArgTypeNameT = TypeVar("ArgTypeNameT", bound=ArgTypeName)
 
-class CLIApp:
+
+class CLIApp(Generic[ArgTypeNameT]):
```

Comment:
I find making this generic just for the sake of testing superfluous:
1. A lot of complex changes in this PR is caused solely by this.
2. The tests could be perfectly fine if they used well known argument types.
3. Tests are leaning towards testing internals of `CLIApp` instead of testing it like a black box.
4. It kinda doesn't make sense by intuition?

Reply:
> I find making this generic just for the sake of testing superfluous:

I am not making this generic for the sake of testing, but in order to decouple CLI logic from Protostar specific code (e.g. `ProtostarArgType`, `ProtostarCLI`), extracting a new module in the future PR and preserving type correctness.

> The tests could be perfectly fine if they used well known argument types.

It's not about tests.

> Tests are leaning towards testing internals of CLIApp instead of testing it like a black box.

Refactoring tests isn't in the scope of this PR.